cg_lims/EPPs/udf/calculate/maf_calculate_volume.py

```diff
@@ -0,0 +1,129 @@
+    occurs at lower concentration levels where standard pipetting volumes are not enough to reach the desired final
+    concentration and final volume."""
+    sample_volume = MINIMUM_TOTAL_VOLUME * FINAL_CONCENTRATION / sample_concentration
+    final_volume = calculate_final_volume(sample_volume, sample_concentration)
+    water_volume = final_volume - sample_volume
+    qc_flag = QC_FAILED
+    return final_volume, water_volume, sample_volume, qc_flag
+
+
+def calculate_volumes(sample_concentration: float) -> Tuple[float, float, float, str]:
+    """Calculates all volumes (final volume, water volume and sample volume) and sets the QC flag."""
+    for sample_volume in PIPETTING_VOLUMES:
+        final_volume = sample_concentration * sample_volume / FINAL_CONCENTRATION
+        water_volume = final_volume - sample_volume
+        if water_volume < MAXIMUM_WATER_VOLUME:
+            qc_flag = QC_PASSED if sample_volume == 3 else QC_FAILED
+            return final_volume, water_volume, sample_volume, qc_flag
+
+
+def calculate_volume(artifacts: List[Artifact]) -> None:
+    """Determines the final volume, water volume, and sample volume"""
+    failed_artifacts = []
+
+    for artifact in artifacts:
+        sample_concentration = get_sample_concentration(artifact)
+        if sample_concentration is None or sample_concentration < FINAL_CONCENTRATION:
+            LOG.warning(
+                f"Sample concentration too low or missing for sample {artifact.samples[0].name}."
+            )
+            failed_artifacts.append(artifact)
+            continue
+        if sample_concentration >= CONCENTRATION_UPPER_LIMIT:
+            LOG.warning(
+                f"Sample concentration too high for sample {artifact.samples[0].name}."
+            )
+            failed_artifacts.append(artifact)
+            continue
+
+        if sample_concentration < 20:
+            (
+                final_volume,
+                water_volume,
+                sample_volume,
+                qc_flag,
+            ) = calculate_volumes_for_low_concentration_samples(sample_concentration)
+        elif sample_concentration >= 20:
+            final_volume, water_volume, sample_volume, qc_flag = calculate_volumes(
+                sample_concentration
+            )
+        else:
+            LOG.warning(
+                f"Could not calculate sample volume for sample {artifact.samples[0].name}."
+            )
+            failed_artifacts.append(artifact)
+            continue
+
+        artifact.qc_flag = qc_flag
+        artifact.udf["Final Volume (uL)"] = final_volume
+        artifact.udf["Volume H2O (ul)"] = water_volume
+        artifact.udf["Volume of sample (ul)"] = sample_volume
+        artifact.put()
+
+    if failed_artifacts:
+        raise LimsError(
+            f"MAF volume calculations failed for {len(failed_artifacts)} samples, "
+            f"{len(artifacts) - len(failed_artifacts)} updates successful. "
+        )
+
```

Comment:
```suggestion
else: 
     message = "Volumes were calculated for all samples."
     LOG.info(message)
     click.echo(message)
```


Something like that.. Its good to inform the user also if everything went well.

Reply:
there's a message like that on lines 125 -127, is that okay or should I move that into this function?